prov/gni/include/gnix_buddy_allocator.h

```diff
@@ -0,0 +1,177 @@
+
+/* Find the index into the free list with block size LEN. */
+#define LIST_INDEX(LEN, MIN) (size_t) (log2((LEN) / (double) (MIN)))
+
+/**
+ * Structure representing a buddy allocator.
+ *
+ * @var base		The base address of the buffer being managed.
+ * @var len		The length of the buffer the buddy allocator is managing.
+ * @var max		The largest chunk of memory that can be allocated.
+ *
+ * @var nlists		The number of free lists.
+ * @var lists		The array of free lists ordered from smallest block size.
+ * at index 0 to largest block size at index nlists - 1.
+ *
+ * @var bitmap		Each bit is 1 if the block is allocated or split,
+ * otherwise the bit is 0.
+ *
+ * @var lock		The buddy alloc handle lock.
+ */
+typedef struct gnix_buddy_alloc_handle {
+	void *base;
+	size_t len;
+	size_t max;
+
+	size_t nlists;
+	struct dlist_entry *lists;
+
+	gnix_bitmap_t bitmap;
+
+	fastlock_t lock;
+} handle_t;
+
+/**
+ * Creates a buddy allocator
+ *
+ * @param[in] base		Base address of buffer to be managed by
+ * allocator.
+ *
+ * @param[in] len		Size of the buffer to be managed by allocator
+ * (must be a multiple of max).
+ *
+ * @param[in] max		Maximum amount of memory that can be allocated
+ * by a single call to _gnix_buddy_alloc (power 2).
+ *
+ * @param[in/out] alloc_handle	Handle to be used for when allocating/freeing
+ * memory managed by the buddy allocator.
+ *
+ * @return FI_SUCCESS		Upon successfully creating an allocator.
+ *
+ * @return -FI_EINVAL		Upon an invalid parameter.
+ *
+ * @return -FI_ENOMEM		Upon failure to allocate memory to create the
+ * buddy allocator.
+ */
+int _gnix_buddy_allocator_create(void *base, size_t len, size_t max,
+				 handle_t **alloc_handle);
+
+/**
+ * Releases all resources associated with a buddy allocator handle.
+ *
+ * @param[in] alloc_handle	Buddy alloc handle to destroy.
+ *
+ * @return FI_SUCCESS	 	Upon successfully destroying an allocator.
+ *
+ * @return -FI_EINVAL 		Upon an invalid parameter.
+ */
+int _gnix_buddy_allocator_destroy(handle_t *alloc_handle);
+
+/**
+ * Allocate a buffer from the buddy allocator
+ *
+ * @param[in] alloc_handle 	Previously allocated GNI buddy_alloc_handle to
+ * use as allocator.
+ *
+ * @param[in/out] ptr		Pointer to an address where the address of the
+ * allocated buffer will be returned.
+ *
+ * @param[in] len		Size of buffer to allocate in bytes.
+ *
+ * @return FI_SUCCESS		Upon successfully allocating a buffer.
+ *
+ * @return -FI_ENOMEM 		Upon not being able to allocate a buffer of the
+ * requested size.
+ *
+ * @return -FI_EINVAL 		Upon an invalid parameters.
+ */
+int _gnix_buddy_alloc(handle_t *alloc_handle, void **ptr, size_t len);
+
+/**
+ * Free a previously allocated buffer
+ *
+ * @param[in] alloc_handle 	Previously allocated GNI buddy_alloc_handle to
+ * use as allocator.
+ *
+ * @param[in/out] ptr		Pointer to an address where the address of the
+ * allocated buffer will be returned.
+ *
+ * @param[in] len		Size of buffer to allocate in bytes.
+ *
```

Comment:
I'd suggest rewording the description of len to something like:
"size of previously allocated buffer in byte"